src/systems/systems/src/air_conditioning/acs_controller.rs

```diff
@@ -0,0 +1,2054 @@
+        }
+        // Single pack operation determination
+        if pack_flow_valve[0].fcv_is_open() != pack_flow_valve[1].fcv_is_open() {
+            intermediate_flow =
+                intermediate_flow.max(Ratio::new::<percent>(Self::ONE_PACK_FLOW_LIMIT));
+        }
+        if matches!(
+            aircraft_state,
+            AirConditioningStateManager::BeginTakeOff(_)
+                | AirConditioningStateManager::EndTakeOff(_)
+                | AirConditioningStateManager::BeginLanding(_)
+                | AirConditioningStateManager::EndLanding(_)
+        ) {
+            intermediate_flow =
+                intermediate_flow.min(Ratio::new::<percent>(Self::FLOW_REDUCTION_LIMIT));
+        }
+        intermediate_flow.max(Ratio::new::<percent>(Self::BACKFLOW_LIMIT))
+    }
+
+    // This calculates the flow based on the demand, when the packs are modelled this needs to be changed
+    // so the demand actuates the valve, and then the flow is calculated based on that
+    fn absolute_flow_calculation(&self, pressurization: &impl Cabin) -> MassRate {
+        let absolute_flow = self.flow_demand.get::<ratio>()
+            * (Self::FLOW_CONSTANT_XCAB * pressurization.altitude().get::<foot>()
+                + Self::FLOW_CONSTANT_C);
+        MassRate::new::<kilogram_per_second>(absolute_flow)
+    }
+
+    fn fcv_open_allowed_determination(
+        &mut self,
+        acs_overhead: &AirConditioningSystemOverhead<ZONES>,
+        engine_fire_push_buttons: &impl EngineFirePushButtons,
+        pressurization_overhead: &PressurizationOverheadPanel,
+        pneumatic: &(impl PneumaticBleed + EngineStartState),
+    ) {
+        // Flow Control Valve 1
+        self.fcv_1_open_allowed = acs_overhead.pack_pushbuttons_state()[0]
+            && !(pneumatic.left_engine_state() == EngineState::Starting)
+            && (!(pneumatic.right_engine_state() == EngineState::Starting)
+                || !pneumatic.engine_crossbleed_is_on())
+            && !engine_fire_push_buttons.is_released(1)
+            && !pressurization_overhead.ditching_is_on();
+        // && ! pack 1 overheat
+        // Flow Control Valve 2
+        self.fcv_2_open_allowed = acs_overhead.pack_pushbuttons_state()[1]
+            && !(pneumatic.right_engine_state() == EngineState::Starting)
+            && (!(pneumatic.left_engine_state() == EngineState::Starting)
+                || !pneumatic.engine_crossbleed_is_on())
+            && !engine_fire_push_buttons.is_released(2)
+            && !pressurization_overhead.ditching_is_on();
+        // && ! pack 2 overheat
+    }
+
+    fn should_open_fcv_determination(
+        &self,
+        engines: [&impl EngineCorrectedN1; 2],
+        pneumatic: &(impl PneumaticBleed + EngineStartState),
+        pneumatic_overhead: [bool; 2],
+    ) -> [bool; 2] {
+        // Pneumatic overhead represents engine bleed pushbutton for left and right engine(s)
```

Comment:
Is there a reason you chose to pass down `pneumatic_overhead` as an array of booleans instead of the actual overhead panel type or a trait implementing the bleed pushbutton states?

Reply:
Yup, mainly because this is in the `systems` crate and the overhead panel type is A320 specific. My reasoning is that the A380 (and I believe any other aircraft type) also has 2 packs but of course with 4 engines the overhead type will be different. By passing booleans we can extract the logic at a higher level, since the system itself is only concerned with whether the pack is receiving bleed air from the engines or not.

Of course when the packs are fully modelled this logic will move from here to the packs themselves.

Reply:
And... as soon as I said this I realised I could get the same result with a trait! Forget what I said, implementing a trait instead.

Reply:
Done 😃